module2/gift-certificate-system/persistence/src/main/java/com/epam/esm/dao/GiftCertificateDao.java

```diff
@@ -0,0 +1,14 @@
+package com.epam.esm.dao;
+
+import com.epam.esm.model.GiftCertificate;
+
+import java.util.List;
+
+public interface GiftCertificateDao extends CrudDAO<GiftCertificate> {
+
+    List<GiftCertificate> getCertificatesByTagName(String tagName);
+    List<GiftCertificate> getCertificatesByPartOfName(String partOfName);
+    List<GiftCertificate> getCertificatesByPartOfDescription(String partOfDescription);
```

Comment:
Join all these methods plus findAll() into one method. Basically they are doing the same thing. You can use basic where clause like _WHERE 1=1_ for default query and append optional conditions programmatically.